event-sourcing/event-store-jpa/src/main/java/org/apache/james/eventsourcing/eventstore/jpa/JPAEventStore.java

```diff
@@ -0,0 +1,176 @@
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.eventsourcing.eventstore.jpa;
+
+
+import static org.apache.james.eventsourcing.eventstore.jpa.model.JPAEvent.DELETE_AGGREGATE_QUERY;
+import static org.apache.james.eventsourcing.eventstore.jpa.model.JPAEvent.SELECT_AGGREGATE_QUERY;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import jakarta.inject.Inject;
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.EntityManagerFactory;
+import jakarta.persistence.EntityTransaction;
+import jakarta.persistence.PersistenceException;
+import jakarta.persistence.PersistenceUnit;
+
+import org.apache.james.backends.jpa.EntityManagerUtils;
+import org.apache.james.eventsourcing.AggregateId;
+import org.apache.james.eventsourcing.Event;
+import org.apache.james.eventsourcing.eventstore.EventStore;
+import org.apache.james.eventsourcing.eventstore.EventStoreFailedException;
+import org.apache.james.eventsourcing.eventstore.History;
+import org.apache.james.eventsourcing.eventstore.JsonEventSerializer;
+import org.apache.james.eventsourcing.eventstore.jpa.model.JPAEvent;
+import org.reactivestreams.Publisher;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.google.common.base.Preconditions;
+
+import reactor.core.publisher.Mono;
+import scala.collection.Iterable;
+
+
+public class JPAEventStore implements EventStore {
+
+    /**
+     * The entity manager to access the database.
+     */
+    private EntityManagerFactory entityManagerFactory;
+
+    /**
+     * The JSON serializer to serialize the event data.
+     */
+    private JsonEventSerializer jsonEventSerializer;
+
+    /**
+     * Set the serializer to use.
+     */
+    @Inject
+    public void setJsonEventSerializer(JsonEventSerializer jsonEventSerializer) {
+        this.jsonEventSerializer = jsonEventSerializer;
+    }
+
+    /**
+     * Set the entity manager to use.
+     */
+    @Inject
+    @PersistenceUnit(unitName = "James")
+    public void setEntityManagerFactory(EntityManagerFactory entityManagerFactory) {
+        this.entityManagerFactory = entityManagerFactory;
+    }
+
+    @Override
+    public Publisher<Void> appendAll(Iterable<Event> events) {
+        if (events.isEmpty()) {
+            return Mono.empty();
+        }
+        Preconditions.checkArgument(Event.belongsToSameAggregate(events));
+        AggregateId aggregateId = events.head().getAggregateId();
+        EntityManager entityManager = entityManagerFactory.createEntityManager();
+        final EntityTransaction transaction = entityManager.getTransaction();
```

Comment:
We can handle this with TransactionRunner I believe

```
return Mono.fromRunnable(() -> TransactionRunner.runAndHandleException(entityManager -> {
             events.foreach(Throwing.runnable(e -> {
                    JPAEvent jpaEvent = new JPAEvent(aggregateId, e.eventId(), jsonEventSerializer.serialize(e));
                    entityManager.persist(jpaEvent);
            }));
}, e -> {
   throw  new EventStoreFailedException("Unable to add events", e);
});
```

Note that:
 - `e` variable name is ambiguous, we shall favor `event`
 - Wrap the whole transformation into a mono not just the final transformation

Reply:
The TransactionWorker api isn't as shown above, but was easy enough to adjust. However the Throwing part I gave a few tries with several variations and types but couldn't get to work yet. If you have a working snippet that could save me some time trying to mess with the java/scala type system incompatibilities (Function1 is not a Function etc)...

btw I think it would have been nice if at least the events source api part remained in java (even if other module implementations were in scala) rather than making the interface itself use scala objects and have the scala types leak into the java code... scala was designed to try an play nice with java, but not necessarily the other way around. but perhaps that's just my own preference or lack of experience stitching them together :-)

Reply:
For now I got it to work with JavaConverters.asJava(events), a bit ugly but I don't know how to get the scala objects to play nice with Throwing. If you have a better way let me know.